sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,358 @@
+    }
+
+    val minString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+    val maxString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.max match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+
+    /**
+     * decide if the value is in [min, max] of the column.
+     * We currently don't store min/max for binary/string type.
+     * Hence, we assume it is in boundary for binary/string type.
+     */
+    val inBoundary: Boolean = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) =>
+            datumString.toDouble >= minString.toDouble && datumString.toDouble <= maxString.toDouble
+          case None => true
+        }
+      case _ => true
+    }
+
+    if (inBoundary) {
+      percent = 1.0 / ndv.toDouble
+
+      if (update) {
+        /**
+         * We update ColumnStat structure after apply this equality predicate.
+         * Set distinctCount to 1.  Set nullCount to 0.
+         */
+        val oneBigInt: BigInt = 1
+        val zeroBigInt: BigInt = 0
+        val newStats = attrRef.dataType match {
+          case dataType: DataType if (dataType.isInstanceOf[NumericType]
+            || dataType.isInstanceOf[DateType]
+            || dataType.isInstanceOf[TimestampType]) =>
+            val newValue = Some(datumString.toDouble)
+            aColStat.copy(distinctCount = oneBigInt, min = newValue,
+              max = newValue, nullCount = zeroBigInt)
+          case _ => aColStat.copy(distinctCount = oneBigInt, nullCount = zeroBigInt)
+        }
+        mutableColStats += (attrRef.name -> newStats)
+      }
+
+    } else {
+      percent = 0.0
+    }
+
+    percent
+  }
+
+  def evaluateInSet(
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      hSet: Set[Any],
+      update: Boolean)
+    : Double = {
+    if (!planStat.colStats.contains(attrRef.name)) {
+      logInfo("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+    // TODO: will fill in this method later.
+    1.0
+  }
+
+  def evaluateBinaryForNumeric(
+      op: BinaryComparison,
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+    // TODO: will fill in this method later.
+    1.0
+  }
+
+  def evaluateBinaryForString(

```

Comment:
I don't think we need to deal with binary predicates for string neither, because now we don't have max/min or equi-width histogram for string.